--- a/include/media/camera_common.h
+++ b/include/media/camera_common.h
@@ -29,6 +29,7 @@
 #include <media/v4l2-subdev.h>
 #include <media/v4l2-ctrls.h>
 #include <media/tegracam_core.h>
+#include <media/gmsl-link.h>
 
 /*
  * Scaling factor for converting a Q10.22 fixed point value
@@ -46,6 +47,11 @@
 	u16 val;
 };
 
+struct reg_32 {
+	u32 addr;
+	u32 val;
+};
+
 struct camera_common_power_rail {
 	struct regulator *dvdd;
 	struct regulator *avdd;
@@ -70,13 +76,16 @@
 	const char *parentclk_name; /* NULL for no parent clock*/
 	unsigned int pwdn_gpio;
 	unsigned int reset_gpio;
+	unsigned int xmaster_gpio;
 	unsigned int af_gpio;
+	char *gmsl;
 	bool ext_reg;
 	int (*power_on)(struct camera_common_power_rail *pw);
 	int (*power_off)(struct camera_common_power_rail *pw);
 	struct camera_common_regulators regulators;
 	bool use_cam_gpio;
 	bool has_eeprom;
+	bool has_color_filter;
 	bool v_flip;
 	bool h_mirror;
 	unsigned int fuse_id_addr;
@@ -105,6 +114,13 @@
 				int num_override_regs,
 				u16 wait_ms_addr, u16 end_addr);
 
+int
+regmap_util_write_table_32(struct regmap *regmap,
+			  const struct reg_32 table[],
+			  const struct reg_32 override_list[],
+			  int num_override_regs,
+			  u32 wait_ms_addr, u32 end_addr);
+
 enum switch_state {
 	SWITCH_OFF,
 	SWITCH_ON,
@@ -160,6 +176,10 @@
 	  u16 addr, u8 val);
 	int (*read_reg)(struct camera_common_data *s_data,
 	  u16 addr, u8 *val);
+	int (*write_reg_32)(struct camera_common_data *s_data,
+	  u32 addr, u32 val);
+	int (*read_reg_32)(struct camera_common_data *s_data,
+	  u32 addr, u32 *val);
 	struct camera_common_pdata *(*parse_dt)(struct tegracam_device *tc_dev);
 	int (*power_get)(struct tegracam_device *tc_dev);
 	int (*power_put)(struct tegracam_device *tc_dev);
@@ -168,6 +188,9 @@
 	int (*set_mode)(struct tegracam_device *tc_dev);
 	int (*start_streaming)(struct tegracam_device *tc_dev);
 	int (*stop_streaming)(struct tegracam_device *tc_dev);
+	int (*check_unsupported_mode)(struct camera_common_data *s_data, struct v4l2_mbus_framefmt *mf);
+	int (*after_set_pixel_format)(struct camera_common_data *s_data);
+	int (*init_private_controls)(struct tegracam_device *tc_dev);
 };
 
 struct tegracam_sensor_data {
@@ -182,6 +205,7 @@
 	const u32 *ctrl_cid_list;
 	bool is_blob_supported;
 	int (*set_gain)(struct tegracam_device *tc_dev, s64 val);
+	int (*set_digital_gain)(struct tegracam_device *tc_dev, s64 val);
 	int (*set_exposure)(struct tegracam_device *tc_dev, s64 val);
 	int (*set_exposure_short)(struct tegracam_device *tc_dev, s64 val);
 	int (*set_frame_rate)(struct tegracam_device *tc_dev, s64 val);
@@ -192,6 +216,24 @@
 				struct v4l2_ctrl *ctrl);
 	int (*fill_compound_ctrl)(struct tegracam_device *tc_dev,
 				struct v4l2_ctrl *ctrl);
+	int (*set_test_pattern)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_data_rate)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_conversion_gain)(struct tegracam_device *tc_dev, bool val);
+	int (*set_adc_mode)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_gain_offset)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_exposure_threshold)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_exposure_threshold_high)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_gc_point_first)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_gc_point_second)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_gc_gain_point_first)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_gc_gain_point_second)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_operation_mode)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_sync_mode)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_broadcast_ctrl)(struct tegracam_device *tc_dev,
+                            struct v4l2_ctrl *ctrl);
+	int (*set_shutter_mode)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_orientation)(struct tegracam_device *tc_dev, u32 val);
+	int (*set_black_level)(struct tegracam_device *tc_dev, s64 val);
 	int (*set_gain_ex)(struct tegracam_device *tc_dev,
 			struct sensor_blob *blob, s64 val);
 	int (*set_exposure_ex)(struct tegracam_device *tc_dev,
@@ -229,21 +271,32 @@
 	/* TODO: cleanup neeeded once all the sensors adapt new framework */
 	struct tegracam_ctrl_handler		*tegracam_ctrl_hdl;
 	struct regmap				*regmap;
+	struct regmap 				*broadcast_regmap;
 	struct camera_common_pdata		*pdata;
+	struct device				*ser_dev;
+	struct device				*dser_dev;
+	struct gmsl_link_ctx			g_ctx;
 	/* TODO: cleanup needed for priv once all the sensors adapt new framework */
 	void	*priv;
+	s64	exposure_min_range, exposure_max_range;
+	s64	short_exposure_min_range, short_exposure_max_range;
 	int	numctrls;
 	int	csi_port;
 	int	numlanes;
 	int	mode;
 	int 	mode_prop_idx;
 	int	numfmts;
+	int 	blklvl_max_range;
+	int 	dig_gain_def_value;
+	int 	dig_gain_max_range;
+	int 	dig_gain_min_range;
 	int	def_mode, def_width, def_height;
 	int	def_clk_freq;
 	int	fmt_width, fmt_height;
 	int	sensor_mode_id;
 	bool	use_sensor_mode_id;
 	bool	override_enable;
+	bool	group_hold_active;
 	u32	version;
 };
 
--- a/include/media/tegra-v4l2-camera.h
+++ b/include/media/tegra-v4l2-camera.h
@@ -62,6 +62,37 @@
 #define TEGRA_CAMERA_CID_VI_PREFERRED_STRIDE (TEGRA_CAMERA_CID_BASE+110)
 #define TEGRA_CAMERA_CID_VI_CAPTURE_TIMEOUT (TEGRA_CAMERA_CID_BASE+111)
 
+#define TEGRA_CAMERA_CID_DATA_RATE                  (TEGRA_CAMERA_CID_BASE+120)
+#define TEGRA_CAMERA_CID_TEST_PATTERN               (TEGRA_CAMERA_CID_BASE+121)
+#define TEGRA_CAMERA_CID_OPERATION_MODE             (TEGRA_CAMERA_CID_BASE+122)
+#define TEGRA_CAMERA_CID_SYNC_MODE                  (TEGRA_CAMERA_CID_BASE+123)
+#define TEGRA_CAMERA_CID_BROADCAST                  (TEGRA_CAMERA_CID_BASE+124)
+#define TEGRA_CAMERA_CID_BLACK_LEVEL                (TEGRA_CAMERA_CID_BASE+125)
+#define TEGRA_CAMERA_CID_HDR_CONTROL                (TEGRA_CAMERA_CID_BASE+126)
+#define TEGRA_CAMERA_CID_HDR_LIN_KNEE1              (TEGRA_CAMERA_CID_BASE+127)
+#define TEGRA_CAMERA_CID_HDR_LIN_KNEE2              (TEGRA_CAMERA_CID_BASE+128)
+#define TEGRA_CAMERA_CID_DIGITAL_GAIN               (TEGRA_CAMERA_CID_BASE+129)
+#define TEGRA_CAMERA_CID_FLIP_MIRROR                (TEGRA_CAMERA_CID_BASE+130)
+#define TEGRA_CAMERA_CID_SHUTTER_MODE               (TEGRA_CAMERA_CID_BASE+131)
+#define TEGRA_CAMERA_CID_XVS_XHS_STATE              (TEGRA_CAMERA_CID_BASE+132)
+#define TEGRA_CAMERA_CID_CONVERSION_GAIN            (TEGRA_CAMERA_CID_BASE+133)
+
+/* Sensor specific controls */
+
+#define TEGRA_CAMERA_CID_PWL_LUT_CONTROL            (TEGRA_CAMERA_CID_BASE+150)
+#define TEGRA_CAMERA_CID_DIGITAL_GAIN_SHORT         (TEGRA_CAMERA_CID_BASE+151)
+#define TEGRA_CAMERA_CID_BLACK_LEVEL_SHORT          (TEGRA_CAMERA_CID_BASE+152)
+
+#define TEGRA_CAMERA_CID_ADC_MODE                   (TEGRA_CAMERA_CID_BASE+220)
+#define TEGRA_CAMERA_CID_GAIN_OFFSET                (TEGRA_CAMERA_CID_BASE+221)
+#define TEGRA_CAMERA_CID_EXPOSURE_THRESHOLD         (TEGRA_CAMERA_CID_BASE+222)
+#define TEGRA_CAMERA_CID_GRADATION_COMPRESSION      (TEGRA_CAMERA_CID_BASE+223)
+#define TEGRA_CAMERA_CID_GC_START_POINT_FIRST       (TEGRA_CAMERA_CID_BASE+224)
+#define TEGRA_CAMERA_CID_GC_START_POINT_SECOND      (TEGRA_CAMERA_CID_BASE+225)
+#define TEGRA_CAMERA_CID_GC_GAIN_POINT_FIRST        (TEGRA_CAMERA_CID_BASE+226)
+#define TEGRA_CAMERA_CID_GC_GAIN_POINT_SECOND       (TEGRA_CAMERA_CID_BASE+227)
+#define TEGRA_CAMERA_CID_EXPOSURE_THRESHOLD_HIGH    (TEGRA_CAMERA_CID_BASE+228)
+
 /**
  * This is temporary with the current v4l2 infrastructure
  * currently discussing with upstream maintainers our proposals and
--- a/include/media/tegra_camera_core.h
+++ b/include/media/tegra_camera_core.h
@@ -18,7 +18,7 @@
 /* Width alignment */
 #define TEGRA_WIDTH_ALIGNMENT	1
 /* Stride alignment */
-#define TEGRA_STRIDE_ALIGNMENT	1
+#define TEGRA_STRIDE_ALIGNMENT	256
 /* Height alignment */
 #define TEGRA_HEIGHT_ALIGNMENT	1
 /* Size alignment */
--- a/include/media/vi.h
+++ b/include/media/vi.h
@@ -60,6 +60,32 @@
 #define NUM_VI_WATCHDOG					2
 #endif
 
+/*
+XXX Beware, these defines are part of a patch that is (hopefully) a temporary implementation of accessing embedded data on jetson tx2.
+It relies on a custom patch in the VI4 driver on Jetson TX2.
+Most likely we will be able to implement a similar patch for Xavier and Nano platforms (VI2 and VI5).
+
+The patch works by allocating some extra memory after each buffer, and the VI is modified to put the embedded data there.
+The buffer zone is used to store the width and height of embedded data (2 bytes each),
+but it has to be 0x100 bytes because that is the smallest increment by which the dma target addresses can be adjusted.
+The embedded data is stored after the buffer "buffer zone".
+
+buffer
++---------------------------------+
+|pixel data                       |
+|                                 |
++---------------------------------+
+|buffer zone of 0x100 bytes       |
++---------------------------------+
+|embedded data                    |
++---------------------------------+
+
+ ASCII art generated using https://asciiflow.com/
+ */
+
+#define EMBEDDED_DATA_MAX_SIZE ((8*64*1024))
+#define EMBEDDED_DATA_BUFFER_ZONE_SIZE ((0x100))
+
 typedef void (*callback)(void *);
 
 struct tegra_vi_stats {
