--- a/drivers/media/i2c/Makefile
+++ b/drivers/media/i2c/Makefile
@@ -3,8 +3,16 @@
 
 subdir-ccflags-y += -Werror
 
+obj-m += fr_common.o
+fr_common-objs := fr_sensor_common.o ./../../i2c/fr_i2c_generic.o
+
+obj-m += fr_lifmd_lvds2mipi_1.o
+obj-m += fr_lifcl_var2fixed_1.o
+
 obj-m += max9295.o
 obj-m += max9296.o
+obj-m += fr_max96792.o
+obj-m += fr_max96793.o
 ifeq ($(findstring ack_src,$(NV_BUILD_KERNEL_OPTIONS)),)
 obj-m += max96712.o
 
@@ -21,6 +29,25 @@
 obj-m += nv_ar0234.o
 obj-m += nv_hawk_owl.o
 obj-m += max929x.o
+obj-m += fr_imx283.o
+obj-m += fr_imx296.o
+obj-m += fr_imx304.o
+obj-m += fr_imx335.o
+obj-m += fr_imx412.o
+obj-m += fr_imx464.o
+obj-m += fr_imx530.o
+obj-m += fr_imx565.o
+obj-m += fr_imx568.o
+obj-m += fr_imx577.o
+obj-m += fr_imx585.o
+obj-m += fr_imx636.o
+obj-m += fr_imx662.o
+obj-m += fr_imx675.o
+obj-m += fr_imx676.o
+obj-m += fr_imx678.o
+obj-m += fr_imx715.o
+obj-m += fr_imx838.o
+obj-m += fr_imx900.o
 endif
 
 obj-m += pca9570.o
